cas-compute/src/symbolic/derivative.rs

```diff
@@ -0,0 +1,268 @@
+use rug::Integer;
+
+use crate::primitive::{float, int};
+
+use super::expr::Primary;
+use super::Expr;
+
+// Something that is not trivially zero - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_zero(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => {
+            !f.is_zero()
+        },
+        Expr::Primary(Primary::Integer(i)) => {
+            !i.is_zero()
+        },
+        Expr::Add(sum) => {
+            sum.iter().any(non_zero)
+        },
+        Expr::Mul(mul) => {
+            mul.iter().all(non_zero)
+        }
+        _ => true
+    }
+}
+
+// Something that is not trivially one - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_unity(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => !f.eq(&float(1)),
+        Expr::Primary(Primary::Integer(i)) => !i.eq(&int(1)),
+        Expr::Mul(exprs) => {
+            exprs.iter().any(non_unity)
+        },
+        Expr::Exp(expr, expr1) => {
+            non_zero(expr1) && non_unity(expr)
+        },
+        _ => true
+    }
+}
+
+#[derive(Default)]
+struct MultBuilder(Vec<Expr>);
```

Comment:
```suggestion
/// Helper struct to build a product of expressions while applying basic simplification rules. If
/// any of the expressions are zero, the product is reduced to zero.
#[derive(Default)]
struct MultBuilder(Vec<Expr>);
```

Document private APIs as well.